storage/scratchpad/src/sparse_merkle/mod.rs

```diff
@@ -228,6 +249,347 @@ where
         Ok((result_hashes, current_state_tree))
     }
 
+    /// Constructs a new Sparse Merkle Tree, returns the SMT root hash after each update and the
+    /// final SMT root. Since the tree is immutable, existing tree remains the same and may
+    /// share parts with the new, returned tree. Unlike `serial_update', intermediate trees aren't
+    /// constructed, but only root hashes are computed. `batches_update' takes value reference
+    /// because the algorithm requires a copy per value at the end of tree traversals. Taking
+    /// in a reference avoids double copy (by the caller and by the implementation).
+    pub fn batches_update(
+        &self,
+        update_batch: Vec<Vec<(HashValue, &V)>>,
+        proof_reader: &impl ProofRead<V>,
+    ) -> Result<(Vec<HashValue>, Self), UpdateError> {
+        if update_batch.is_empty() {
+            // No updates.
+            return Ok((vec![], self.clone()));
+        }
+        
+        // Construct (key, txn_id, value) update vector, where 0 <= txn_id < update_batch.len().
+        // The entries are sorted and deduplicated (keeping last) based on the (key, txn_id) pair.
+        let updates: Vec<(HashValue, usize, &V)> = update_batch.iter()
```

Comment:
into_iter?

Reply:
done! couldn't get rid of move because it's w.r.t. the i which is out of the scope of the inner closure